core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```
@@ -194,21 +214,32 @@ class GroupCoordinator(val brokerId: Int,
       } else if (!group.supportsProtocols(protocolType, MemberMetadata.plainProtocolSet(protocols))) {
         responseCallback(joinError(memberId, Errors.INCONSISTENT_GROUP_PROTOCOL))
       } else if (group.isPendingMember(memberId)) {
-        // A rejoining pending member will be accepted.
-        addMemberAndRebalance(rebalanceTimeoutMs, sessionTimeoutMs, memberId, clientId, clientHost, protocolType,
-          protocols, group, responseCallback)
+        // A rejoining pending member will be accepted. Note that pending member will never be a static member.
+        assert(groupInstanceId == JoinGroupRequest.EMPTY_GROUP_INSTANCE_ID)
```

Do we want to use assert here? It will crash the broker if this happens.

---

The reason to use `assert` is to prevent future implementation from breaking the existing assumption. Basically known static member should never be `pending`.

---

We generally frown on assertions. It is usually better to raise an exception with a clear message.

---

Could you share more details? The reason for using assertion is to avoid creating invalid state from the code change stage. For example, we have
```private def onGroupLoaded(group: GroupMetadata) {
    group.inLock {
      info(s"Loading group metadata for ${group.groupId} with generation ${group.generationId}")
      assert(group.is(Stable) || group.is(Empty))
      if (groupIsOverCapacity(group)) {
        prepareRebalance(group, s"Freshly-loaded group is over capacity ($groupConfig.groupMaxSize). Rebalacing in order to give a chance for consumers to commit offsets")
      }

      group.allMemberMetadata.foreach(completeAndScheduleNextHeartbeatExpiration(group, _))
    }```
and ```
 // trigger the awaiting join group response callback for all the members after rebalancing
          for (member <- group.allMemberMetadata) {
            assert(member.awaitingJoinCallback != null)
            val joinResult = JoinGroupResult(
              members = if (group.isLeader(member.memberId)) {
```
It would be great if you shed light on the trade-offs on these cases, thank you!

---

@hachikuji Could you give me some guidance on this? Thank you!

---

Yeah to be honest we do use assertions somewhere like @abbccdda mentioned in onGroupLoaded; they are used to indicate "this should never happen, and if it happens, it's a bug". As a hind-sight we can actually just replace with if-throw-illegal-state-exception across the board so that when it happens indeed, it will crash hard but leave us a meaning stack trace.

---

@guozhangwang I see, so we should choose to throw exception for most times?

---

Yeah I'd suggest so.

---

Yes, exactly. We like stack traces and nice error messages! A lot of these impossible states have a way of becoming more possible over time.